### ISSUE TITLE
datetime or duration (e.g. on x axis) results in error

### ISSUE THREAD

**Issue by erikhuizinga** [matlab2tikz/matlab2tikz#963]
Based on the [comment](https://github.com/matlab2tikz/matlab2tikz/issues/383#issuecomment-260650219) by [awesomeindustries](https://github.com/awesomeindustries) in [issue 383](https://github.com/matlab2tikz/matlab2tikz/issues/383), the following code results in an error in MALTAB R2016b.

```matlab
data = rand(10, 1);
plotdates = datetime : datetime+numel(data)-1;
figure
plot(plotdates, data)
matlab2tikz file
```

I, too, have this problem on MATLAB R2016b. I do not have R2016a or earlier installed any more, so I cannot test if this is new behaviour or that it has existed before. Issue 383 was closed as solved in November 2015, so this issue could be present in R2016a too, assuming the closer had no problems on R2015b.

The error text:

```matlab
Error using datetime/horzcat (line 1260)
All inputs must be datetimes or date/time character vectors or date/time strings.
Error in matlab2tikz>getXYZDataFromLine (line 1829)
        data = [xData(:), yData(:)];
Error in matlab2tikz>drawLine (line 1735)
    data       = getXYZDataFromLine(m2t, h);
Error in matlab2tikz>handleAllChildren (line 681)
                [m2t, str] = drawLine(m2t, child);
Error in matlab2tikz>drawAxes (line 880)
    [m2t, childrenEnvs] = handleAllChildren(m2t, handle);
Error in matlab2tikz>saveToFile (line 482)
        m2t = drawAxes(m2t, relevantAxesHandle);
Error in matlab2tikz (line 334)
    m2t = saveToFile(m2t, fid, fileWasOpen); 
```

It seems `matlab2tikz>getXYZDataFromLine (line 1829)
        data = [xData(:), yData(:)];` gets `datetime` data and non-`datetime` data and tries to concatenate the two, which isn't allowed for these types of data (`double` and `datetime`). EDIT: it appears to be allowed for `double` and `duration`, e.g. try `[1 seconds]`. However, the error if you issue `[1 datetime]` is the same as the one `matlab2tikz` throws:

```matlab
Error using datetime/horzcat (line 1260)
All inputs must be datetimes or date/time character vectors or date/time strings. 
```

This may be fixable by converting the `data` matrix implementation into a `cell` or `table`. With `table`, a lot more is possible too, so this may be preferable. However, it was introduced in R2013b, which is not too long ago. so a `cell` may be a better type to improve backwards compatibility. It seems not too difficult to adjust the other functions using `data` to work with a `table` or `cell`. However, I can't tell how to properly handle MATLAB's built-in types other than the numeric types, i.e. `datetime` and `duration`, further down the line.

Using a `duration` results in different errors. Example code:

```matlab
data = rand(10,1);
plotdates = seconds(1:10);
figure
plot(plotdates,data)
matlab2tikz file
```
Error output:
```matlab
Error using sprintf
Unable to convert 'duration' value to 'double'.
Error in matlab2tikz>makeTable (line 5391)
        table{iRow} = sprintf(FORMAT, thisData{:});
Error in matlab2tikz>plotLine2d (line 1866)
    [m2t, table, tableOptions] = makeTable(m2t, repmat({''}, size(data,2)), data);
Error in matlab2tikz>writePlotData (line 1806)
            [m2t, Part] = plotLine2d(m2t, opts, dataCell{k});
Error in matlab2tikz>drawLine (line 1744)
    [m2t, dataString]  = writePlotData(m2t, data, drawOptions);
Error in matlab2tikz>handleAllChildren (line 681)
                [m2t, str] = drawLine(m2t, child);
Error in matlab2tikz>drawAxes (line 880)
    [m2t, childrenEnvs] = handleAllChildren(m2t, handle);
Error in matlab2tikz>saveToFile (line 482)
        m2t = drawAxes(m2t, relevantAxesHandle);
Error in matlab2tikz (line 334)
    m2t = saveToFile(m2t, fid, fileWasOpen); 
```
This time the conversion of `duration` to `double` isn't allowed. A `duration` has the difficulty it is dependent on units of time. A `duration` in seconds is different from a `duration` in hours. `matlab2tikz` could just use any unit to determine the coordinates of the TikZ plot, but then the final tick labels should respect the MATLAB plot's tick label format (I assume `matlab2tikz` already does this) or take the `Format` property of the `datetime` or `duration` to create the appropriate labels.

**Comment by egeerardyn** [comment]
Thank you for this complete bug report. 

I indeed agree with your analysis. Probably using a cell array (over table, though that truely is a nice data type) is best for the reasons you point out (backwards compatibility), properly handling this may ripple through to a very large part of our code base as we use that pattern of putting all data into a big matrix quite often (that makes our life easy when working with `hgtransform`, which is also a feature that some people tend to use).

Another option would be to investigate casting those durations/datetimes/... into doubles one way or another and only at the extremities of m2t handle the conversion between dates/times and doubles.

**Comment by erikhuizinga** [comment]
I think the latter option may be best while noting that, whenever new data types or specialised plot objects are introduced, the first option could be better, but much more work to implement.

The latter option would require the recognition of `datetime` or `duration` input on any axis, set a flag for later down the line and convert into `double` indeed. The conversion should indeed take place near the extremities.

I guess the implementation will require minor additions to the 'recognise & convert' part and quite a bit more work on the LaTeX output part, as dates, times and durations are, in general, very challenging to work with due to the numerous notations, standards, etc.

I also guess there must already be some awesome date and time packages for LaTeX that could be made a requirement for end users when any input contains `datetime` or `duration` data. Such a package should support many, if not all notations MATLAB's [date and time format](https://mathworks.com/help/matlab/ref/datetime-properties.html#property_Format) and [duration formats](https://mathworks.com/help/matlab/ref/duration.html#input_argument_displayFormat) support.

**Comment by erikhuizinga** [comment]
Side note: the newly introduces `timeseries` data type has a plot method. It results in a plot with `datetime` type data on the horizontal axis.

**Comment by okomarov** [comment]
This should have been fixed by PR #964, so closing.

**Comment by SimGas** [comment]
I still feature problems when using datetime on a normal xy-plot.
Minimum (not)working example:
`time= [datetime(2018,12,31,12,0,0):minutes(1):datetime(2018,12,31,12,10,0)];
data= [0:1:10];

plot(time,data)
matlab2tikz('testplot.tex')`
results in:

> Error using datetime/horzcat (line 1286)
All inputs must be datetimes or date/time character vectors or date/time strings.

Error in matlab2tikz>getXYZDataFromLine (line 1829)
        data = [xData(:), yData(:)];

Error in matlab2tikz>drawLine (line 1735)
    data       = getXYZDataFromLine(m2t, h);

Error in matlab2tikz>handleAllChildren (line 681)
                [m2t, str] = drawLine(m2t, child);

Error in matlab2tikz>drawAxes (line 880)
    [m2t, childrenEnvs] = handleAllChildren(m2t, handle);

Error in matlab2tikz>saveToFile (line 482)
        m2t = drawAxes(m2t, relevantAxesHandle);

Error in matlab2tikz (line 334)
    m2t = saveToFile(m2t, fid, fileWasOpen);
